q2_feature_table/_split.py

```diff
@@ -0,0 +1,51 @@
+# ----------------------------------------------------------------------------
+# Copyright (c) 2016-2023, QIIME 2 development team.
+#
+# Distributed under the terms of the Modified BSD License.
+#
+# The full license is in the file LICENSE, distributed with this software.
+# ----------------------------------------------------------------------------
+
+import biom
+import qiime2
+
+
+_valid_value_characters = set(
+    "0123456789"
+    "abcdefghijklmnopqrstuvwxyz"
+    "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
+    "-_.")
+
+
+def _check_valid_value(value):
+    return len(set(value) - _valid_value_characters) != 0
+
+
+def _check_valid_values(values):
+    invalid_values = [v for v in values if _check_valid_value(v)]
+    if len(invalid_values) > 0:
+        raise ValueError(
+            "Invalid value(s) identified during feature table splitting. "
+            "Values in metadata columns used for splitting can only contain "
+            "alphanumeric characters, dashes (-), dots (.), and underscores "
+            "(_). The problematic value(s) are: "
+            f"{', '.join(invalid_values)}")
+
+
+def split(table: biom.Table,
+          metadata: qiime2.CategoricalMetadataColumn,
+          filter_empty_features: bool = True) -> biom.Table:
+    metadata_df = metadata.drop_missing_values().to_dataframe()
+
+    indices = metadata_df.reset_index(
+        ).groupby(metadata.name)[metadata_df.index.name].apply(list).to_dict()
+
+    _check_valid_values(indices.keys())
+
+    result = {}
+    for group, sample_ids in indices.items():
+        t = table.filter(sample_ids, axis='sample', inplace=False)
+        if filter_empty_features:
+            t.remove_empty(axis='observation', inplace=True)
+        result[group] = t
```

Comment:
I think this could be expressed as the following:

```suggestion
    lookup = {}
    for group, sample_ids in indices.items():
        for sid in sample_ids:
            lookup[sid] = group
            
    def group_f(i, m):
        return lookup[i]
        
    result = dict(table.partition(group_f))
```

...I suspect the construction of `lookup` could be done directly from Pandas as well.

For reference, see [`Table.partition`](https://github.com/biocore/biom-format/blob/941973a6aac3bd2621be8818e9e15ccb9e63bdb2/biom/table.py#L2398-L2446)

Reply:
Thanks @wasade - I forgot about the `Table.partition` method.

Reply:
I tried this approach, but it doesn't handle the case where there is missing metadata for sample(s), in which case we'd not include those sample(s) in any of the resulting tables (i.e., my `SplitTests.test_missing_data_samples_dropped`). Is there a way to handle that with `Table.partition`? 

Since we're getting close to release time we can always circle back to this in future.

Reply:
Ah good catch. One possibility is to dump them into an "__unsupported__" bin or some unlikely to exist key, and then remove it from the dict after? I agree that would be a helpful API expansion -- I'll open an issue on BIOM

Reply:
Thanks for creating the issue @wasade! I'm going to hold off on using Table.partition here until that feature is added, but will happily delegate all of this to Table.partition then. I made a couple of additional suggestions on that issue that would move all of the interesting bits of this function to the `Table.partition` method (though those suggestions aren't blockers for using this here, just conveniences).